<a href="https://colab.research.google.com/github/ChakritKammason/DPDM/blob/master/Association_Rules.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
import os
base_path = '/content/drive/My Drive/DWDM_Material'
os.listdir(base_path)

Mounted at /content/drive


['teach.csv',
 '03 สังคม.xls',
 '52835793_259306464964511_3851804852342489088_n.png',
 'Copy of search_click.csv',
 'Copy of w_restaurant.csv',
 'metadata.xlsx',
 'Preprocessing02.html',
 'reduced_marketbasket.csv',
 'search_click.csv',
 'teach02.csv',
 'test.csv',
 'w_restaurant.csv',
 'AnimalsPlague.csv',
 '666d790734bee7737c0b5144c462b9b5.jpg',
 'AnimalPlague.csv',
 'AnimalsCorn.csv',
 'Local.xlsx',
 'Local.csv',
 'AnimalCorn.csv',
 'AnimalCorn.xlsx']

In [2]:
data = pd.read_csv(os.path.join(base_path,'reduced_marketbasket.csv'),encoding = 'unicode_escape')
data

InvoiceNo StockCode  ... CustomerID         Country
0         536365    85123A  ...    17850.0  United Kingdom
1         536365     71053  ...    17850.0  United Kingdom
2         536365    84406B  ...    17850.0  United Kingdom
3         536365    84029G  ...    17850.0  United Kingdom
4         536365    84029E  ...    17850.0  United Kingdom
...          ...       ...  ...        ...             ...
541904    581587     22613  ...    12680.0          France
541905    581587     22899  ...    12680.0          France
541906    581587     23254  ...    12680.0          France
541907    581587     23255  ...    12680.0          France
541908    581587     22138  ...    12680.0          France

[541909 rows x 8 columns]

In [9]:
data.groupby('InvoiceNo').sum()

Quantity  UnitPrice  CustomerID
InvoiceNo                                 
536365           40      27.37    124950.0
536366           12       3.70     35700.0
536367           83      58.24    156564.0
536368           15      19.10     52188.0
536369            3       5.95     13047.0
...             ...        ...         ...
C581484      -80995       2.08     16446.0
C581490         -23       2.78     28794.0
C581499          -1     224.69     15498.0
C581568          -5      10.95     15311.0
C581569          -6       2.50     34630.0

[25900 rows x 3 columns]

In [10]:
transacs = data.groupby('InvoiceNo')['StockCode'].apply(list)
transacs

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

# Apriori

http://zaxrosenberg.com/unofficial-apyori-documentation/

In [5]:
!pip install apyori # แพ็คเก็จที่คนไม่ค่อยใช้ต้องทำแบบนี้ก่อนทุกครั้งก่อนใช้งาน

  Created wheel for apyori: filename=apyori-1.1.2-cp36-none-any.whl size=5975 sha256=36b7bf3eac16e8fc7c3e143f639ad94f7f50e3fb41972390d52371d1aabf22f2
  Stored in directory: /root/.cache/pip/wheels/5d/92/bb/474bbadbc8c0062b9eb168f69982a0443263f8ab1711a8cad0
Successfully built apyori


In [6]:
from apyori import apriori

In [11]:
out = list(apriori(transacs, min_support = 0.03, min_confidence = 0.1))
out

[RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)]),
 RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])]

In [12]:
out[0]

RelationRecord(items=frozenset({'22386', '85099B'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])

Quiz หา k-itemset ที่มีความน่าสนใจ (โดยพิจารณาลูกค้าเป็นรายคน) พร้อมทั้งอธิบายว่าน่าสนใจยังไง

In [13]:
transacs = data.groupby('CustomerID')['StockCode'].apply(list)
transacs

CustomerID
12346.0                                       [23166, 23166]
12347.0    [85116, 22375, 71477, 22492, 22771, 22772, 227...
12348.0    [84992, 22951, 84991, 84991, 21213, 21213, 226...
12349.0    [23112, 23460, 21564, 21411, 21563, 22131, 221...
12350.0    [21908, 22412, 79066K, 79191C, 22348, 84086C, ...
                                 ...                        
18280.0    [82484, 22180, 22467, 22725, 22727, 22495, 223...
18281.0    [22037, 22716, 22028, 23007, 23008, 23209, 22467]
18282.0    [21270, 23187, 23295, 22089, 21108, 21109, 224...
18283.0    [22356, 20726, 22384, 22386, 20717, 20718, 850...
18287.0    [22755, 22754, 22753, 22756, 22758, 22757, 227...
Name: StockCode, Length: 4372, dtype: object

In [15]:
out = list(apriori(transacs, min_support = 0.1, min_confidence = 0.2))
out

[RelationRecord(items=frozenset({'22423'}), support=0.20288197621225984, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.20288197621225984, lift=1.0)])]

อาจารย์ทำ

In [16]:
out = list(apriori(transacs, min_support = 0.07, min_confidence = 0.7))
out

[RelationRecord(items=frozenset({'85123A', '21733'}), support=0.07136322049405307, ordered_statistics=[OrderedStatistic(items_base=frozenset({'21733'}), items_add=frozenset({'85123A'}), confidence=0.8103896103896104, lift=4.129397874852421)]),
 RelationRecord(items=frozenset({'22699', '22423'}), support=0.07410795974382434, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22423'}), confidence=0.7587822014051523, lift=3.740017795426523)]),
 RelationRecord(items=frozenset({'22699', '22697'}), support=0.07410795974382434, ordered_statistics=[OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.8307692307692307, lift=8.506143037290578), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7587822014051523, lift=8.506143037290578)])]

In [17]:
transacs = data.groupby('CustomerID')['Description'].apply(list)
transacs

CustomerID
12346.0    [MEDIUM CERAMIC TOP STORAGE JAR, MEDIUM CERAMI...
12347.0    [BLACK CANDELABRA T-LIGHT HOLDER, AIRLINE BAG ...
12348.0    [72 SWEETHEART FAIRY CAKE CASES, 60 CAKE CASES...
12349.0    [PARISIENNE CURIO CABINET, SWEETHEART WALL TID...
12350.0    [CHOCOLATE THIS WAY METAL SIGN, METAL SIGN NEI...
                                 ...                        
18280.0    [WOOD BLACK BOARD ANT WHITE FINISH, RETROSPOT ...
18281.0    [ROBOT BIRTHDAY CARD, CARD CIRCUS PARADE, PENN...
18282.0    [ANTIQUE CREAM CUTLERY CUPBOARD, FRENCH STYLE ...
18283.0    [CHARLOTTE BAG PINK POLKADOT, LUNCH BAG WOODLA...
18287.0    [SMALL PURPLE BABUSHKA NOTEBOOK , SMALL RED BA...
Name: Description, Length: 4372, dtype: object

In [18]:
out = list(apriori(transacs, min_support = 0.07, min_confidence = 0.7))
out

[RelationRecord(items=frozenset({'GREEN REGENCY TEACUP AND SAUCER', 'ROSES REGENCY TEACUP AND SAUCER '}), support=0.07410795974382434, ordered_statistics=[OrderedStatistic(items_base=frozenset({'GREEN REGENCY TEACUP AND SAUCER'}), items_add=frozenset({'ROSES REGENCY TEACUP AND SAUCER '}), confidence=0.8307692307692307, lift=8.506143037290578), OrderedStatistic(items_base=frozenset({'ROSES REGENCY TEACUP AND SAUCER '}), items_add=frozenset({'GREEN REGENCY TEACUP AND SAUCER'}), confidence=0.7587822014051523, lift=8.506143037290578)]),
 RelationRecord(items=frozenset({'WHITE HANGING HEART T-LIGHT HOLDER', 'RED HANGING HEART T-LIGHT HOLDER'}), support=0.07136322049405307, ordered_statistics=[OrderedStatistic(items_base=frozenset({'RED HANGING HEART T-LIGHT HOLDER'}), items_add=frozenset({'WHITE HANGING HEART T-LIGHT HOLDER'}), confidence=0.8103896103896104, lift=4.129397874852421)]),
 RelationRecord(items=frozenset({'ROSES REGENCY TEACUP AND SAUCER ', 'REGENCY CAKESTAND 3 TIER'}), support=

In [19]:
ID = list(apriori(transacs, min_support = 0.03, min_confidence = 0.1))
ID

[RelationRecord(items=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), support=0.15530649588289114, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), confidence=0.15530649588289114, lift=1.0)]),
 RelationRecord(items=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), support=0.1338060384263495, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), confidence=0.1338060384263495, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER LARGE'}), support=0.11139066788655078, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER LARGE'}), confidence=0.11139066788655078, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER SMALL'}), support=0.13106129917657822, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER SMALL'}), confidence=0.13106129917657822, lift=1.0)])

In [20]:
ID_2=list(apriori(transacs, min_support = 0.04, min_confidence = 0.1))
ID_2

[RelationRecord(items=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), support=0.15530649588289114, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), confidence=0.15530649588289114, lift=1.0)]),
 RelationRecord(items=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), support=0.1338060384263495, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), confidence=0.1338060384263495, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER LARGE'}), support=0.11139066788655078, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER LARGE'}), confidence=0.11139066788655078, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER SMALL'}), support=0.13106129917657822, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER SMALL'}), confidence=0.13106129917657822, lift=1.0)])

In [21]:
ID_3=list(apriori(transacs, min_support = 0.05, min_confidence = 0.1))
ID_3

[RelationRecord(items=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), support=0.15530649588289114, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), confidence=0.15530649588289114, lift=1.0)]),
 RelationRecord(items=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), support=0.1338060384263495, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), confidence=0.1338060384263495, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER LARGE'}), support=0.11139066788655078, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER LARGE'}), confidence=0.11139066788655078, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER SMALL'}), support=0.13106129917657822, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER SMALL'}), confidence=0.13106129917657822, lift=1.0)])

In [22]:
ID_4=list(apriori(transacs, min_support = 0.06, min_confidence = 0.1))
ID_4

[RelationRecord(items=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), support=0.15530649588289114, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), confidence=0.15530649588289114, lift=1.0)]),
 RelationRecord(items=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), support=0.1338060384263495, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), confidence=0.1338060384263495, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER LARGE'}), support=0.11139066788655078, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER LARGE'}), confidence=0.11139066788655078, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER SMALL'}), support=0.13106129917657822, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER SMALL'}), confidence=0.13106129917657822, lift=1.0)])

In [23]:
ID_5=list(apriori(transacs, min_support = 0.07, min_confidence = 0.1))
ID_5

[RelationRecord(items=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), support=0.15530649588289114, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'ASSORTED COLOUR BIRD ORNAMENT'}), confidence=0.15530649588289114, lift=1.0)]),
 RelationRecord(items=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), support=0.1338060384263495, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'BAKING SET 9 PIECE RETROSPOT '}), confidence=0.1338060384263495, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER LARGE'}), support=0.11139066788655078, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER LARGE'}), confidence=0.11139066788655078, lift=1.0)]),
 RelationRecord(items=frozenset({'HEART OF WICKER SMALL'}), support=0.13106129917657822, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'HEART OF WICKER SMALL'}), confidence=0.13106129917657822, lift=1.0)])

In [24]:
ID_6=list(apriori(transacs, min_support = 0.07, min_confidence = 0.7))
ID_6

[RelationRecord(items=frozenset({'GREEN REGENCY TEACUP AND SAUCER', 'ROSES REGENCY TEACUP AND SAUCER '}), support=0.07410795974382434, ordered_statistics=[OrderedStatistic(items_base=frozenset({'GREEN REGENCY TEACUP AND SAUCER'}), items_add=frozenset({'ROSES REGENCY TEACUP AND SAUCER '}), confidence=0.8307692307692307, lift=8.506143037290578), OrderedStatistic(items_base=frozenset({'ROSES REGENCY TEACUP AND SAUCER '}), items_add=frozenset({'GREEN REGENCY TEACUP AND SAUCER'}), confidence=0.7587822014051523, lift=8.506143037290578)]),
 RelationRecord(items=frozenset({'WHITE HANGING HEART T-LIGHT HOLDER', 'RED HANGING HEART T-LIGHT HOLDER'}), support=0.07136322049405307, ordered_statistics=[OrderedStatistic(items_base=frozenset({'RED HANGING HEART T-LIGHT HOLDER'}), items_add=frozenset({'WHITE HANGING HEART T-LIGHT HOLDER'}), confidence=0.8103896103896104, lift=4.129397874852421)]),
 RelationRecord(items=frozenset({'ROSES REGENCY TEACUP AND SAUCER ', 'REGENCY CAKESTAND 3 TIER'}), support=


เนื่องจากคนที่จะซื้อสินค้า {'21733', '85123A'} มีความน่าจะเป็นคือ 0.07136322049405307

เนื่องจากคนที่จะซื้อสินค้า {'22423', '22699'} มีความน่าจะเป็นคือ 0.07410795974382434

เนื่องจากคนที่จะซื้อสินค้า {'22697', '22699'} มีความน่าจะเป็นคือ 0.07410795974382434

และ lift > 1 ทุกคู่ ดังนั้นจะมีโปรโมชั่นให้ ถ้าลูกค้าซื้อสินค้าคู่ดังกล่าวสะสมครบ 10 ครั้ง ทางร้านจะแถมสินค้า 1 คู่ให้ โดยลูกค้าสามารถเลือกคู่ใดคู่หนึ่งได้ฟรี